### ISSUE TITLE
pp_ctl.c: new build-time warning on non-debugging builds

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#21393]
A new build-time warning has begun to be emitted from `pp_ctl.c` in **non-debugging** builds.  On Ubuntu Linux 22.04 LTS, using `gcc-11` as the C-compiler, I configured with:
```
sh ./Configure -des -Dusedevel
```
... and built `perl` at b78c98a142.  I recorded the following warning during `make test_prep`:
```
pp_ctl.c: In function ‘S_require_file’:
pp_ctl.c:4454:13: warning: unused variable ‘count’ [-Wunused-variable]
 4454 |         int count = call_sv(PL_hook__require__before, G_SCALAR);
      |             ^~~~~
```
I recorded the same warning when building with `clang14` on FreeBSD-13 in a threaded build.

The statement at which the warning is being emitted was added in the previous development cycle:
```
commit 93f6f9654a81b66c47c07ca982e4c00558bd4159
Author:     Yves Orton <demerphq@gmail.com>
AuthorDate: Mon Dec 19 19:32:03 2022 +0100
Commit:     Yves Orton <demerphq@gmail.com>
CommitDate: Sat Mar 18 20:57:59 2023 +0800

    pp_ctl.c - add support for hooking require.
```
However, until recently it was not emitting any warning!  Further investigation showed that there were recent changes in `pp_ctl.c` just after the statement in question.  Specifically, in 5abffa7ff8a, committed to blead on Aug 16, we see:
```
commit 5abffa7ff8a6cee3e13e061f8332cf567427a21c
Author:     David Mitchell <davem@iabyn.nospamdeletethisbit.com>
AuthorDate: Tue Feb 14 12:57:30 2023 +0000
Commit:     David Mitchell <davem@iabyn.nospamdeletethisbit.com>
CommitDate: Wed Aug 16 17:16:59 2023 +0100

    make RC-stack-aware: require,eval
    
    Update pp_entereval() et al to handle a reference-counted stack
    environment in the presence of PERL_RC_STACK.
    
    But in the presence of PERL_XXX_TMP_NORC, don't actually manipulate
    reference counts yet. This will be turned off in a few commits' time.

diff --git a/pp_ctl.c b/pp_ctl.c
index 62f2b8a8f0..488d337fde 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c

...

@@ -4383,15 +4388,17 @@ S_require_file(pTHX_ SV *sv)

...
         int count = call_sv(PL_hook__require__before, G_SCALAR);
-        SPAGAIN;
-        if (count && SvOK(*SP) && SvROK(*SP) && SvTYPE(SvRV(*SP)) == SVt_PVCV)
-            /* the RC++ preserves it across the FREETMPS below */
-            post_hook__require__before_sv = SvREFCNT_inc_simple_NN(*SP);
+        SV *rsv = *PL_stack_sp;
+        assert(count == 1); /* scalar context */
```
So `count` was removed at the line containing `count && SvOK(*SP) && ...`, but added in the line with `assert`.

I reconfigured with the following:
```
sh ./Configure -des -Dusedevel -DDEBUGGING
```
... then ran `make test_prep` again (with `gcc-11` again as the compiler).  This time, no warning at all was emitted.  So the regular build gives a warning; the debugging build does not.

@iabyn, could you investigate this and see if there is a way to eliminate the warning during non-debugging builds?

Thank you very much.

**Comment by leonerd** [comment]
Having a quick look now - it appears the `count` variable is only ever used during an `assert()`, so indeed it would always warn on non-debugging builds.

I'm not sure if we have a standard way around that. I suppose you could just pop in a `PERL_UNUSED_VAR(count)` but a quick hunt through the code doesn't reveal that to be common practice.

**Comment by jkeenan** [comment]
Should be fixed by #21408
